netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -423,7 +839,479 @@ net_ebpf_extension_sock_addr_authorize_connection_classify(
+            sock_addr_ctx,
+            incoming_metadata_values->transportEndpointHandle,
+            true, /* original */
+            v4_mapped);
+        _net_ebpf_ext_connection_context_initialize_key(
+            &connection_context_redirected->key,
+            sock_addr_ctx,
+            incoming_metadata_values->transportEndpointHandle,
+            false, /* original */
+            v4_mapped);
+
+        verdict = BPF_SOCK_ADDR_VERDICT_PROCEED;
+        redirected = true;
+
+        goto CreateContext;
+    }
+
+    // Acquire classify handle.
+    status = FwpsAcquireClassifyHandle((void*)classify_context, 0, &classify_handle);
+    if (!NT_SUCCESS(status)) {
+        NET_EBPF_EXT_LOG_NTSTATUS_API_FAILURE_UINT64_UINT64(
+            NET_EBPF_EXT_TRACELOG_KEYWORD_SOCK_ADDR,
+            "FwpsAcquireClassifyHandle",
+            status,
+            filter->filterId,
+            (uint64_t)sock_addr_ctx->compartment_id);
+
+        goto Exit;
+    }
+    classify_handle_acquired = true;
+
+    if (v4_mapped) {
+        sock_addr_ctx->family = AF_INET;
+        const uint8_t* v4_ip = IN6_GET_ADDR_V4MAPPED((IN6_ADDR*)&sock_addr_ctx->user_ip6);
+        uint32_t local_v4_ip = *((uint32_t*)v4_ip);
+        memset(sock_addr_ctx->user_ip6, 0, 16);
+        sock_addr_ctx->user_ip4 = local_v4_ip;
+    }
+
+    compartment_id = filter_context->compartment_id;
+    ASSERT((compartment_id == UNSPECIFIED_COMPARTMENT_ID) || (compartment_id == sock_addr_ctx->compartment_id));
+    if (compartment_id != UNSPECIFIED_COMPARTMENT_ID && compartment_id != sock_addr_ctx->compartment_id) {
+        // The client is not interested in this compartment Id. Change action to PERMIT.
+        NET_EBPF_EXT_LOG_MESSAGE_UINT32(
+            NET_EBPF_EXT_TRACELOG_LEVEL_VERBOSE,
+            NET_EBPF_EXT_TRACELOG_KEYWORD_SOCK_ADDR,
+            "The cgroup_sock_addr eBPF program is not interested in this compartment ID",
+            sock_addr_ctx->compartment_id);
+
+        action = FWP_ACTION_PERMIT;
+        goto Exit;
+    }
+
+    _net_ebpf_ext_connection_context_initialize_key(
+        &connection_context_original->key,
+        sock_addr_ctx,
+        incoming_metadata_values->transportEndpointHandle,
+        true,
+        v4_mapped);
+
+    if (net_ebpf_extension_hook_invoke_program(attached_client, sock_addr_ctx, &verdict) != EBPF_SUCCESS) {
+        status = STATUS_UNSUCCESSFUL;
+        goto Exit;
+    }
+
+    // Initialize connection_context_redirected destination with the redirected address.
+    _net_ebpf_ext_connection_context_initialize_key(
+        &connection_context_redirected->key,
+        sock_addr_ctx,
+        incoming_metadata_values->transportEndpointHandle,
+        false,
+        v4_mapped);
+
+    if (v4_mapped) {
+        sock_addr_ctx->family = AF_INET6;
+        IN_ADDR v4_address = *((IN_ADDR*)&sock_addr_ctx->user_ip4);
+        IN6_SET_ADDR_V4MAPPED((IN6_ADDR*)&sock_addr_ctx->user_ip6, (IN_ADDR*)&v4_address);
```

Comment:
Why not do this prior to creating the rediected context key. And all this should happen inside process_redirect_verdict IMO.

Reply:
This is reverting the modification done at line 1197.
In case of v4-mapped addresses, redirected context key needs to be populated with v4 address, and not v4-mapped v6 address (as AUTH layer will get v4 addresses). Hence redirected context is populated before converting back to v4-mapped v6 addresses.